### PR TITLE
[2/2] Use quorum_id for leader election and flexiraft watermark

### PR DESCRIPTION
**Pull request by li-chi**
This PR currently includes both first part of changes and second part of changes. You can start with PR167 and ignore the duplicated in this PR for easier review. You can also review both changes as a whole in this PR if you like.

# Sub-region commit

When commit use_quorum_id is true, quorum is defined by quorum_id, where the MySQL instance and its two LBUs should have the same quorum_id. In peer attribute, each peer has the unchanged region property, plus the new quorum_id.  Commit watermark calculation and leader election calculation is based on quorum_id instead of region. 

## Changes in this PR
1) Whenever attr().region() is used, replaced with a generic function that return region or quorum_id based on commit rule. (Except for proxy and RegionDurableIndex)
2) Add support to edit local_peer's quorum_id as well as tracked peer's quorum_id in consensus queue.

## Tests have done
All the tests from previous PR and
1. Conversion between use_quorum_id
2. Reject append entries on all members, except for the leader its one LBU In a 6-voter region (frc in stage-1 unittestdb), observe commit marker still moving. (commit sub-region is working)
3. Move leadership to a 3-voter region, withhold votes on one frc mysql and its two lbus, and start election on the other frc mysql instance. Observe that frc instance wins election and is able to commit normally. (candidate sub-region is working)

### CODE REVIEW

File: src/kudu/consensus/raft_consensus.cc
Comment (li-chi) [inline src/kudu/consensus/raft_consensus.cc]: This check is not required when quorum_id is used.

File: src/kudu/consensus/raft_consensus.cc
Comment (alanliang) [inline src/kudu/consensus/raft_consensus.cc]: s/ChangeQuorumTYpe/ChangeQuorumType

File: src/kudu/consensus/raft_consensus.h
Comment (alanliang) [inline src/kudu/consensus/raft_consensus.h]: ditto

File: src/kudu/consensus/quorum_util.cc
Comment (li-chi) [inline src/kudu/consensus/quorum_util.cc:949]: GetRegionalCountsFromConfig: Rename to GetQuorumIdlCountsFromConfig

File: src/kudu/consensus/raft_consensus.cc
Comment (li-chi) [inline src/kudu/consensus/raft_consensus.cc]: Abstract this function into quorum_util (peer, type) -> string

File: src/kudu/consensus/leader_election.cc
Comment (li-chi) [inline src/kudu/consensus/leader_election.cc]: Have one map, key can be quorum_id or region

File: src/kudu/consensus/quorum_util.h
Comment (li-chi) [inline src/kudu/consensus/quorum_util.h:51]: This function is still needed. It is called by MysqlRaft.cpp, and used for master ban, which is still based on region instead of quorum_id

File: src/kudu/consensus/quorum_util.cc
Comment (li-chi) [inline src/kudu/consensus/quorum_util.cc]: If not exist map

File: src/kudu/consensus/consensus_queue.cc
Comment (li-chi) [inline src/kudu/consensus/consensus_queue.cc]: Remain check(). Have a separate function for update.

File: src/kudu/consensus/consensus_queue.cc
Comment (li-chi) [inline src/kudu/consensus/consensus_queue.cc]: quorum id

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc]: As discussed in 1:!, Why are we updating peer quorum id each time SetLeaderMode is called or TrackedPeer handling is happening. setting of quorum id seems unrelated to the original intent of function CheckPeersInActiveConfigIfLeaderUnlocked

File: src/kudu/consensus/raft_consensus.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:3817]: non-voters don't require quorum id, right?

**Review by anirbanr-fb — COMMENTED** [review]
the stack of commits seems to have the same commit twice. 
""[1/2] Allow changing peer quorum_id and allow changing use_quorum_id
 "" - is showing up twice

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc:1216]: i think this CHECK is still valid? We want region property to be always present?

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc]: nit: leader_quorum
 nit: watermarks_in_leader_quorum

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc]: nit: peer_quorum

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc:1258]: this check is still valid, I think although it might be happening at more appropriate places

File: src/kudu/consensus/leader_election.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: this is copying the reference from uuid_to_quorum_id_ into a local map for each call.
we can eliminate this expensive copy each time.

File: src/kudu/consensus/leader_election.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: this has no protection for optional field quorum_id

File: src/kudu/consensus/leader_election.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: creates a local copy of the map

File: src/kudu/consensus/leader_election.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: nit: variable names should change to quorum

File: src/kudu/consensus/leader_election.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: correctly passing by reference, but none of the callers are using it.

File: src/kudu/consensus/leader_election.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: change variable names to be consistent with "quorum" instead of quorum

Comment (anirbanr-fb) [reply src/kudu/consensus/leader_election.cc]: Modify the comments as well to be reflective of change

File: src/kudu/consensus/quorum_util.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/quorum_util.cc]: return variable should be quorum, not region

File: src/kudu/consensus/quorum_util.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/quorum_util.cc]: leader_quorum

File: src/kudu/consensus/quorum_util.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/quorum_util.cc:898]: "quorum"

File: src/kudu/consensus/consensus_queue.cc
Comment (li-chi) [reply src/kudu/consensus/consensus_queue.cc:1216]: getQuorumId() has a check on it when Region is used:
`CHECK(peer.has_attrs() && peer.attrs().has_region());`, which should be good enough?

Feel like a bit strange to have a check on region here, as we want region/quorum_id to be hidden by GetQuorumId() in this file.

File: src/kudu/consensus/consensus_queue.cc
Comment (li-chi) [reply src/kudu/consensus/consensus_queue.cc:1258]: Same above. GetQuorumId() has a check on peer_pb.has_attrs() and region()/quorum_id().

I can keep the CHECK(peer_pb.has_permanent_uuid()) here.

File: src/kudu/consensus/leader_election.cc
Comment (li-chi) [reply src/kudu/consensus/leader_election.cc]: Probably came from an older version.

Currently is `uuid_to_quorum_id_.emplace(peer.permanent_uuid(), GetQuorumId(peer, use_quorum_id));`

GetQuorumId() has a CHEKC() on quorum_id.

File: src/kudu/consensus/raft_consensus.cc
Comment (li-chi) [reply src/kudu/consensus/raft_consensus.cc:3817]: yes.

File: src/kudu/consensus/leader_election.cc
Comment (li-chi) [reply src/kudu/consensus/leader_election.cc]: Code is gone. There is only one local field map.

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc:1240]: I am concerned about lines 1240 - 1246. It seems to me we are directly
reading voter_distribution. Will this voter distribution be region based on
quorum id based. During rollouts this seems to me the same watermarking
code can use both types?

Are we recomputing voter_distribution during setting time as we initially discussed

File: src/kudu/consensus/raft_consensus.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:3798]: I put my concerns about voter_distribution in the other commit.
Neither ChangeQuorumType or SetPeerQuorumIds is altering the existing
voter distribution. This is causing a skew between some maps being based
on quorum id and some maps being based on voter distribution and we are
reading it directly in Watermarking code. How are we going to handle this

File: src/kudu/consensus/consensus_queue.cc
Comment (li-chi) [reply src/kudu/consensus/consensus_queue.cc:1240]: Sorry. This is a careless mistake from me. In my very recent change, I reverted the changes in Static Mode so that Static Mode always uses region and original vd. I ended up mistakenly reverting VD in the Dynamic mode rather than Static Mode.

File: src/kudu/consensus/raft_consensus.cc
Comment (li-chi) [reply src/kudu/consensus/raft_consensus.cc:3798]: GetVoterDistributionForQuorumId() in quorum_util.cc is the function to handle this. In QuorumId type, it set vd default to 3, and overrides by VD provided by users. In Region type, it directly uses VD provided by users. In dynamic mode, it is supposed to call this function.

File: src/kudu/consensus/quorum_util.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/quorum_util.cc:969]: we need a comment here which says that even if there is a SKEW between voter distribution and quorum id usage during rollouts, this loop should not alter the voter count map, because quorum id will have a proper prefix.

File: src/kudu/consensus/quorum_util.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/quorum_util.cc:41]: we need to add a comment, here that this is about how META runs its quorums

File: src/kudu/consensus/quorum_util.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/quorum_util.cc]: lets remove this conditional and make sure non-voters have no quorum ID.
Returning empty string is fine I think

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc:1288]: i think we will have a hard time debugging quorums if we  name them quorum_1, quorum_2 etc. We need to print the region as well for easy debugging

Comment (li-chi) [reply src/kudu/consensus/consensus_queue.cc:1288]: To debug quorum_id easier:
(1) We will have a status var to show the actual VD for quorum_id: rpl_raft_voter_distribution_for_quorum_id, which states all the available quorum_id. (diff ready in fbcode)
(2) Either raft-dba status or dba status should print quorum_id besides instance host:port, along with opid, voter_type, etc.
(3) Yes. Automation can uses atn-1, atn-2 as quorum_id, as long as we differentiate enough from data center name.

File: src/kudu/consensus/quorum_util.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/quorum_util.cc:956]: nit: using continue pattern is easier to reason about and saves extra levels of indentation.
(!peer.has_member_type() || peer.member_type() != VOTER)
   continue;

**Review by anirbanr-fb — APPROVED** [review]
I will keep looking but mostly looks good.
I have not reviewed leader_election.cc fully.Stamping to unblock,
but we should wait for current hotfixes to rollout before landing in main branch.

**Review by anirbanr-fb — APPROVED** [review]

File: src/kudu/consensus/raft_consensus.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc]: nit: add comment. map returned is from quorum_id -> #

File: src/kudu/consensus/raft_consensus.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:2385]: potentially add a quorum id based test for this or modify existing test to work
in quorum_id mode as well

File: src/kudu/consensus/raft_consensus.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/raft_consensus.cc:2393]: same comment as some other place. Might make sense to print region as well.
Alt idea is to build quorum_id strings in automation to use the region as well, since for the
forseeable future we will have quorum ids within a region.

e.g. quorum_reg_N
but the N can definitely be global index. the _reg_ is mainly for debugging/logging

File: src/kudu/consensus/consensus_queue.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/consensus_queue.cc]: nit: confusing to have so many getQuorumId() with different cases.
This case also is inconsistent with rest of file.
I think it is fine to use GetQuorumId or change it to something GetQuorumIdUsingCommitRule

File: src/kudu/consensus/leader_election.cc
Comment (anirbanr-fb) [inline src/kudu/consensus/leader_election.cc]: nit: next_leader_quorum_ids

**Pull request merged**
